8264127: ListCell editing status is true, when index changes while editing

FlorianKirmaier opened the pull request:
Fixing ListCell editing status is true, when index changes while editing.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264127](https://bugs.openjdk.java.net/browse/JDK-8264127): ListCell editing status is true, when index changes while editing


### Reviewers
 * [Jeanette Winzenburg](https://openjdk.java.net/census#fastegal) (@kleopatra - Committer)
 * [Ajit Ghaisas](https://openjdk.java.net/census#aghaisas) (@aghaisas - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/441/head:pull/441` \
`$ git checkout pull/441`

Update a local copy of the PR: \
`$ git checkout pull/441` \
`$ git pull https://git.openjdk.java.net/jfx pull/441/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 441`

View PR using the GUI difftool: \
`$ git pr show -t 441`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/441.diff">https://git.openjdk.java.net/jfx/pull/441.diff</a>

</details>

kleopatra commented:
good catch, yet another property that's not correctly updated on index change :) 

Quick questions:
- what are the macroscopic effects (that is in a running list) that you see without fixing it?
- do we want mere cell re-use fire a editCancel? (which the list seems to ignore in the test, don't know why)

FlorianKirmaier commented:
To clarify, this with happening with ListViews in production code.
In some conditions, the status of a single cell doesn't update. The effect is, that a Cell is stuck in the Editing mode.
If I remember correctly, it was irreversible stuck. It happened rarely, it might be connected to cells with variable sizes.

Handling the logic from the ListView seems wrong to me, I think the ListCell should update its state automatically dependent of the properties. But I wouldn't know where to add it in the ListView. If that's how it's done in the other cell-based components, then it would make sense and the code more linear. I'm not really opinionated on how to solve it, just went for the simplest fix I've found.

kleopatra commented:
> 
> Handling the logic from the ListView seems wrong to me, 

looks like I was unclear, because that's a 100% me-too :) 

Reformulating my second sentence in test snippets:

     A:
     cell.updateIndex(1);
     list.edit(1);
     cell.updateIndex(0);
     // failing/passing before/after fix
     assertFalse("cell re-use must update cell editing state" , cell.isEditing());

     B:
    List<EditEvent> events = new ArrayList<EditEvent>();
    list.setOnEditCancel(e -> {
        events.add(e);
    });
     .... setup test state as in A
    // passing/failing before/after fix
    assertTrue("cell re-use must not trigger edit events", events.isEmpty()); 
   
    C:
     .... setup test state as in A
     // assert passing/passing before/after fix     
     assertEquals("cell re-use must not change list editing state", 1, list.getEditingIndex);
    
My question was, whether we agree on B. My wondering was about C passing in the light of B failing.

FlorianKirmaier commented:
I don't think B is right.
I would expect a editCancel event when the index of an editing cell is changed.
If there would be another cell, which will get the index 1 (which isn't the case in this artifical test) then i would expect another editStart event.

On the perspective of the ListView, the index never changes, but it's more relevant that some of the cells cancel/start the editing. If one would want to avoid the cancelEdit, when the index is not changing, then the solution should be to make sure, the index of editing cells never changes, which isn't the scope of this PR.

kleopatra commented:
> 
> 
> I don't think B is right.
> I would expect a editCancel event when the index of an editing cell is changed.
> If there would be another cell, which will get the index 1 (which isn't the case in this artifical test) then i would expect another editStart event.

well, I see your point but think it's arguable :)

- editEvent not/firing on cell re-use is not really specified (even overall edit spec is .. suboptimal, see [my oldish summary](https://github.com/kleopatra/swingempire-fx/wiki/CellEditEvents)) - without that spec, we might feel free to fire 
- from the perspective of edit handlers, that editCancel (and the assumed editStart - suspect it doesn't happen, didn't test though) are just spurious events - they are not really interested in volatile state changes, just introduced by an implementation detail like cell re-use. 

As to the scope of this: as I understand it, its goal is to keep the cell's editing state (flag and visual state) in sync with listview's editingIndex when updating the cell index. That should include both directions

- index == editingIndex -> index != editingIndex
- index != editingIndex -> index == editingIndex

The first is already handled (modulo our disagreement on editEvents :), the second is not.

FlorianKirmaier commented:
@kleopatra 
I've added another test for the case, when index changes in such a way, that the cell should no longer be in the editing state,
and the test seems to work. Do I miss something?

About the Events, when i think about it, I was a bit of biased so I don't have to change anything. So I don't really have an opinion about it. But the PR should only fix the case, which is obviously wrong, and i don't really want to change anything in the event logic.

kleopatra commented:
> @kleopatra
> I've added another test for the case, when index changes in such a way, that the cell should no longer be in the editing state,
> and the test seems to work. Do I miss something?

hmm .. that's weird:  your test (cell 1 -> listEdit 0 -> cell 0) indeed passes, doing it the other way round (cell 0 -> listEdit 1 -> cell 1) fails

    @Test
    public void testChangeCellIndex0ToListEditingIndex1() {
        assertChangeIndexToEditing(0, 1);
    }
    
    @Test
    public void testChangeCellIndex1ToListEditingIndex0() {
        assertChangeIndexToEditing(1, 0);
    }
   
    private void assertChangeIndexToEditing(int initialCellIndex, int listEditingIndex) {
        list.setEditable(true);
        cell.updateListView(list);
        cell.updateIndex(initialCellIndex);
        list.edit(listEditingIndex);
        assertEquals("sanity: list editingIndex ", listEditingIndex, list.getEditingIndex());
        assertFalse("sanity: cell must not be editing", cell.isEditing());
        cell.updateIndex(listEditingIndex);
        assertEquals("sanity: index updated ", listEditingIndex, cell.getIndex());
        assertEquals("list editingIndex unchanged by cell", listEditingIndex, list.getEditingIndex());
        assertTrue(cell.isEditing());
    }
    
Something obvious that I missed or a bummer in the test setup or what else? Any ideas?

FlorianKirmaier commented:
I've looked into it. 
This is somehow linked to the fact, that the default focusIndex for the list is 0 instead of -1.
If you change the updateDefaultFocus methods in ListView, then it works.

The tests are somehow synthetic and don't reflect the real world.
So it's probably ok to leave it as it is. A better test which reflects the real world would be better, but that's probably way harder. 
Then we would have to get the ListCells from the ListView ... Which would basically be reasonable for all of these tests, but i don't think it should be done as part of this PR.

kleopatra commented:
> 
> 
> I've looked into it.
> This is somehow linked to the fact, that the default focusIndex for the list is 0 instead of -1.
> If you change the updateDefaultFocus methods in ListView, then it works.
> 

good dig :) Wouldn't have expected the focus state to have any effect .. so we should make certain it's part of the test setup, f.i. by always setting the focusedIndex (of the model) to -1 (or whatever is required to make all pass) always and document why it's done.

> The tests are somehow synthetic and don't reflect the real world.
 
.. true, nevertheless they souldn't fail/pass on some magic number ;)

And yet another failing test, when changing cell index from -1 -> listEditingIndex:

    @Test
    public void testChangeCellIndexMinusOneToListEditingIndex0() {
        assertChangeIndexToEditing(-1, 1);
    }

This looks not fixable by pre-setting focusIndex ..

Coming back to our disagreement about event firing: I'm still weary with those intermediate events from simple re-use - but found a precedence in [TableCell](https://bugs.openjdk.java.net/browse/JDK-8150525): the same broken cell editing state led to data corruption and was fixed by calling updateEditing (including firing). So doing the same here looks okay after all. Should be documented though, at least by tests (for cancel, B in my very first comment with flipped assertion, similar for start edit).

FlorianKirmaier commented:
I've updated the Fix and the tests.
It contains now the 3 cases you've mentioned and all are passing.
The test from 0 works with setting the initial focus.
The test from -1 works, after updating how the fix works. 
Now the index is updated, after updating the item.

FlorianKirmaier commented:
I've now expanded the tests, to check for the correct number of events, similar to your case B - just with reverted conditions, like you've suggested.

kleopatra commented:
- there's still a failing transition (and not covered by the latest test ;) `cellIndex == editingIndex` to `cellIndex == -1`
- as to your fix: don't see a difference between calling `updateEditing `before/after `updateFocus` - simply moving it after the if block seems to solve all toEditing issues

That said: I'm aware that getting this right is tedious and a bit hard, not because it's rocket science but because we need to cover the corner cases as completely as possible (and often fail in doing so, me included :) Chances are that if we detect a problem with a transition "-1 -> 1", there's a similar problem in the invers "1 -> -1". Even if there isn't, we should make sure there isn't by adding a test. That's why I personally prefer writing dedicated parameterized tests (vs. adding to the existing general purpose test), here in cell/editingIndex. Don't know if you noticed, yesterday I attached such a test to the issue (had to write it for [JDK-8265206](https://bugs.openjdk.java.net/browse/JDK-8265206) anyway, so no big deal to replace TableCell with ListCell) - whether you take it or adjust yours accordingly doesn't matter, as long as the coverage is complete.

BTW, in TableCell the remaining issue (editingIndex -> -1) seems to be related to updateEditing backing out if the current index is -1. It probably should cleanup if in editing state (didn't try yet).

kleopatra commented:
curious: do you understand the dependency of update sequence (updateEditing relative to updateFocus)? 

What I see is

- `focusIndex == cellIndex` and calling updateEditing before updateFocus - all toEditing transitions are failing due to corrupting list editing state (its editingIndex seems is changed to -1, that is effectively canceling the edit)
- `focusIndex == editingIndex` and calling updateEditing after updateFocus - all offEditing transitions are failing due to corrupting list editing state

TableCell does the latter and has no dependency on focus state - at least none I can see ;)

kleopatra commented:
> 
> 
> curious: do you understand the dependency of update sequence (updateEditing relative to updateFocus)?
> 
>  [... ]
> 
>     * `focusIndex == editingIndex` and calling updateEditing after updateFocus - all offEditing transitions are failing due to corrupting list editing state
> 
> 
> TableCell does the latter and has no dependency on focus state - at least none I can see ;)

actually, TableCell behavior only _appears_ to not depend on focus state .. because the cell is never moved into focused state if its TableRow is null ;) With a setup including updateTableRow, its focused state is updated to false when moving off the editing row which triggers the listener in cell which cancels the edit ... so we basically have the same behavior as in ListCell. 

To keep the tests focused on editing (vs a mix-in of other properties), we should probably (?) setup focus to -1 always.

FlorianKirmaier commented:
Finally found time to update the PR.

 1. As suggested, I've moved the updateIndex method below the if.
 2. The test from 1 to -1 is a bit more complicated. Because -1 represents "not editing" the test would be differnt. I've adapted the test to check for "not editing" in the case of -1. With this change the test is green.

 Honestly im not really aware of the Ordering of the methods. I didn't spend a lot of time understanding the inner works of all the controls - im usually more focused on fixing memory leaks.
 
 About the initial focus, i see 3 options:
  - set it to -1 in testChangeIndexToEditing1_jdk_8264127
  - set it to -1 in assertChangeIndexToEditing
  - set it to -1 in the setup method.
  All 3 options work. Im fine with keeping it at the concrete test, because it's the only test who needs it, but im fine with any of the 3 options.

FlorianKirmaier commented:
I've fixed the case similar as the fix in the TreeCell!

Option 2 for Keeping focus on -1 is also implemented (seems like it was done in the previous commit)

kleopatra commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/ListCellTest.java:
```diff
@@ -833,5 +836,79 @@ public void testListCellHeights() {
+            assertTrue("cell re-use must trigger edit events", events.size() == 1);
+        } else {
+            // -1 represents "not editing" for the listview and "no index" for the list cell.
+            assertTrue(!cell.isEditing());
```
same as above

kleopatra commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/ListCellTest.java:
```diff
@@ -833,5 +836,79 @@ public void testListCellHeights() {
+        } else {
+            // -1 represents "not editing" for the listview and "no index" for the list cell.
+            assertTrue(!cell.isEditing());
+            assertTrue("cell re-use must trigger edit events", events.size() == 0);
```
same as above

kleopatra commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/ListCellTest.java:
```diff
@@ -833,5 +836,79 @@ public void testListCellHeights() {
+        list.edit(editingIndex);
+        // here we are certain that the cell is in editing state
+        assertTrue("sanity: cell must be editing", cell.isEditing());
+        cell.updateIndex(-1); // change cell index to negative
+        assertFalse("cell must not be editing if cell index is " + cell.getIndex(), cell.isEditing());
```
missing: assert that a cancelEvent is fired and the list editing state is unchanged (they fail until updateEditing is fixed)

kleopatra commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/ListCellTest.java:
```diff
@@ -833,5 +836,79 @@ public void testListCellHeights() {
+        cell.updateIndex(1);
+        list.edit(1);
+        cell.updateIndex(0);
+        assertTrue(!cell.isEditing());
```
let's not confuse the code reader: `assertTrue(!boolean)` should be `assertFalse(boolean)`

kleopatra reviewed (changes requested):
Fix and tests are not yet quite complete - see my inline comments :)

A general suggestion for the tests: I would prefer to have them consistent in both toEditing and offEditing (or whileEditing)

The transitions in the direction `cell-not-editing` to `cell-editing` are factored into a common method `assertChangeIndexToEditing(...)` and have test methods calling that common method with varied concrete indices (f.i `0 ->1`, `-1 -> 1` etc). The other way round - `cell-editing` to `cell-not-editing` - should be factored in a similar way, with the same index pairs. The advantages, IMO: guaranteed same setup and test completeness for both directions and easy read of the variations

FlorianKirmaier commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/ListCellTest.java:
```diff
@@ -833,5 +836,79 @@ public void testListCellHeights() {
+    }
+
+    @Test
+    public void testChangeIndexToEditing3_jdk_8264127() {
+        assertChangeIndexToEditing(1, -1);
```
I guess it doesn't make sense. Probably missunderstanding somewhere. Should i remove? This would also get rid of the if in the test.

kleopatra replied:
yes, remove it: the focus of this issue is that the cell sync's its own editing state on updateIndex off/to list editing index - I think that implies the list always editing with a valid editingIndex :)

kleopatra replied:
yes, remove it: the focus of this issue is that the cell sync's its own editing state on updateIndex off/to list editing index - I think that implies the list always editing with a valid editingIndex :)

*grrr - why was this not added to my review? Really don't understand this tool .. seems to hate me ..

FlorianKirmaier replied:
Now it's added 3 times. : P The if is now removed.

FlorianKirmaier commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/ListCellTest.java:
```diff
@@ -833,5 +836,79 @@ public void testListCellHeights() {
+        list.setOnEditCancel(e -> {
+            events.add(e);
+        });
+        list.setEditable(true);
```
I've removed this test in favor of the test-method below.

FlorianKirmaier commented on modules/javafx.controls/src/main/java/javafx/scene/control/ListCell.java:
```diff
@@ -541,7 +542,11 @@ private void updateEditing() {
+            if(isEditing()) {
+                cancelEdit();
+            }
+        } else {
```
done

FlorianKirmaier commented:
I now use method for both cases with various indizes. 
I've removed the initial test, and now I'm only using the reusable methods.
The cases are now also symmetric, except the outlinder which causes the "if" in the first test. `assertChangeIndexToEditing(-1, 0); `. Should i remove it?

kleopatra commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/ListCellTest.java:
```diff
@@ -833,5 +836,88 @@ public void testListCellHeights() {
+        // here we are certain that the cell is in editing state
+        assertTrue("sanity: cell must be editing", cell.isEditing());
+        cell.updateIndex(cancelIndex); // change cell index to negative
+        assertFalse("cell must not be editing if cell index is " + cell.getIndex(), cell.isEditing());
+        assertEquals(1, events.size());
+    }
```
still missing: assert that list editingIndex is unchanged (see the assertTo test). Without that assert, you don't see any difference in the test between your previous and this commit (which was surrounding the call to cancelEdit with setting the updateEditingIndex flags), do you ;) 

A personal note: tests are our friends, red-green (repeatedly) the color sequence that brings satisfaction :) Just changing a code snippet without making certain to have a failing test before (and passes after) a change is .. well .. suboptimal.

FlorianKirmaier replied:
Yes, i can confirm that it's required to find the issue with the updateEditingIndex. I've added more tests and in general its now more symmetric to the other test!

kleopatra replied:
missing test for the try/finally

kleopatra commented on modules/javafx.controls/src/main/java/javafx/scene/control/ListCell.java:
```diff
@@ -540,8 +541,13 @@ private void updateEditing() {
         final boolean editing = isEditing();
 
-        // Check that the list is specified, and my index is not -1
-        if (index != -1 && list != null) {
+        if (index == -1 || list == null) {
+            if(editing) {
+                updateEditingIndex = false;
+                cancelEdit();
+                updateEditingIndex = true;
+            }
+        } else {
```
exactly :) And now we have duplicated code blocks in a single method which can  be improved by extracting them. It's a bit a matter of judgement/style of the codebase: in this particular case we probably should because 

- it's the same across all cell implementations
- a slight issue turned up in [review](https://github.com/openjdk/jfx/pull/473#discussion_r617762020) of the exact same code in TableCell (should guarantee that the flag is reliably reset to true, no matter what happens in cancelEdit) - fixing that increases the length of the duplicated block. 

Whatever the final outcome, it should be consistent across cells. We might wait with this until the TableCell is finalized, then do the same here - what do you think?

Unrelated to extracting or not: the sequence should have the same code comment everywhere it appears.

FlorianKirmaier replied:
I've now rewritten the code. It's now way simpler and avoids duplicate code. I guess the different implementation in the different cells all have some subtle differences.

Then i guess we wait for the other review to finish. The change to use try finalize seems very very reasonable and important to me.

kleopatra replied:
> 
> 
> I guess the different implementation in the different cells all have some subtle differences.
> 

actually, I doubt that they are different - my guess: we could formally prove that the logic is equivalent :) Or the other way: if they are, we should understand the differences. On first look your variant seems okay, as well.

> Then i guess we wait for the other review to finish. The change to use try finalize seems very very reasonable and important to me.

so we agree - will wait with a final review until the other is decided. You might consider removing the debug sysouts before that :)

FlorianKirmaier replied:
Hm okay, when we assume the result of match in TreeCell is the same as index != editIndex, then they are identical.
And when we assume that cancelEdit doesn't have relevant side effects, which can easily happen due to various listeners ...

I've removed the accidentally commited print statements!

kleopatra reviewed (changes requested):
fix and tests are very near ready, good work :) 

- fix has some details to finalize pending [review](https://github.com/openjdk/jfx/pull/473#discussion_r617762020) for the exact same in TableCell
- test missing one last assertion (about list editing state)

left inline comments

FlorianKirmaier commented:
I've now added the try-finally statement to reset the updateEditingIndex, just like in the other PR.

kleopatra commented on modules/javafx.controls/src/main/java/javafx/scene/control/ListCell.java:
```diff
@@ -540,24 +541,27 @@ private void updateEditing() {
-                // so that subclasses which override cancelEdit can execute. So,
-                // I have to use a kind of hacky flag workaround.
+
+        if (editing && (index == -1 || list == null || index != editIndex)) {
+            // If my index is not the one being edited then I need to cancel
```
probably just me not being able to see at a glance if this if the overall logic in the method is the exact same as before the fix (modulo the fix itself :), but: I would prefer an extra method used in both early return for the -1/null-list and the old else-if. 

If the second reviewer sees the equivalence at a glance, I'll believe his eyes :)

FlorianKirmaier replied:
I've simplified the if-else statements. Should now be obvious that it's the same as in TreeTableCell.

kleopatra reviewed (changes requested):
coming nearer :) see my inline comments

kleopatra commented on modules/javafx.controls/src/main/java/javafx/scene/control/ListCell.java:
```diff
@@ -338,6 +338,7 @@ public void changed(
             updateSelection();
             updateFocus();
         }
+        updateEditing();
     }
```
forgot yesterday: to keep consistent with TreeTableCell (and TreeCell), the updateEditing should be moved into the else-block (as last call) - couldn't find any difference (in fact couldn't reproduce the error the if/else is supposed to solve) as long as updateEditing behaves, though.

FlorianKirmaier replied:
I've moved it to the if-else. Probably doesn't matter.

FlorianKirmaier commented:
I've added a test for the try-catch-finally block, based on the test in the TreeTableCell. (for some reason i can't comment it in the CR)

kleopatra commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/ListCellTest.java:
```diff
@@ -833,5 +836,124 @@ public void testListCellHeights() {
+        cell.updateIndex(editingIndex);
+        list.edit(editingIndex);
+        assertTrue("sanity: ", cell.isEditing());
+        try {
+            list.edit(intermediate);
+        } catch (Exception ex) {
+            // just catching to test in finally
+        } finally {
+            assertFalse("cell must not be editing", cell.isEditing());
+            assertEquals("table must be editing at intermediate index", intermediate, list.getEditingIndex());
+        }
```
that's good - except for a minor omission: the catch-block is never reached, instead it is printed to sysout (or syserr, don't nail me) That happens because ExpressionHelper kind of swallows all exceptions in change notifications by passing them to the uncaughtExceptionHandler. To fix this, a test should un-/install (in After/Before methods) an uncaught exception handler. For examples, see f.i. TableCellTest.

really minor: replace "table" by "list" in the failure message :)

FlorianKirmaier replied:
I've now added the exception handler, like in TreeCellTest.

And I've fixed the texts of the failure messages.

kleopatra commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/ListCellTest.java:
```diff
@@ -833,5 +836,124 @@ public void testListCellHeights() {
+            // just catching to test in finally
+        } finally {
+            assertFalse("cell must not be editing", cell.isEditing());
+            assertEquals("table editing must be cancelled by cell", notEditingIndex, list.getEditingIndex());
```
same minor c&p error as above

kleopatra reviewed (changes requested):
fix looks okay, commented the added test method inline

kleopatra reviewed (approved):
fix and tests look good :)

kleopatra commented:
wondering why the bot removed the rfr label (about three weeks ago) and re-added it a earlier today?

aghaisas reviewed (commented):
Fix and tests look good to me.
I have listed minor editing comments.

FlorianKirmaier commented:
I've done the 3 requested minor code convention changes

kevinrushforth commented:
This one should have been set to require 2 reviewers. I'll do it now, but even if Skara doesn't remove the `ready` label, please wait to integrate it until @kleopatra re-reviews it.

/reviewers 2

FlorianKirmaier commented:
/integrate

kleopatra commented:
/sponsor
